docs: add reporter configuration ref and document HEARTBEAT_PERIOD behaviour (#314)
kubernetes-sigs/node-readiness-controller: docs/book/src/reference/reporter-configuration.md

# Reporter Configuration

> 

Reviewer: Also add an one-liner at start to describe the component, to set a premise on "what" is node-condition-reporter. It could be conveyed to the user that it is one of the different ways of reporting node conditions from the node. 

In the future we would evolve the docs to include more options such as https://github.com/kubernetes-sigs/node-readiness-controller/pull/318

Author: Noted and updated!

The `readiness-condition-reporter` is configured entirely through environment
variables on its container. This page lists all supported variables. For a
walkthrough of deploying the reporter, see the [CNI Readiness](../examples/cni-readiness.md)

> and [Security Agent](../examples/security-agent-readiness.md) examples.

Reviewer: ```suggestion
This page lists configuration variables for deploying the `readiness-condition-reporter`. These can be configured through environment
variables on its container. See the [CNI Readiness](../examples/cni-readiness.md)
and [Security Agent](../examples/security-agent-readiness.md) for examples.
```

Author: Updated

## Environment Variables

| Variable | Description | Default |
| --- | --- | --- |
| `NODE_NAME` | The name of the Node this reporter instance updates. Typically set via the Downward API (`fieldRef: spec.nodeName`) rather than a literal value. | (required) |
| `CHECK_ENDPOINT` | The local HTTP endpoint the reporter polls to determine health, e.g. `http://localhost:9099/readiness`. | (required) |

> | `CONDITION_TYPE` | The Node condition type the reporter writes, e.g. `projectcalico.org/CalicoReady`. | (required) |

Reviewer: ```suggestion
| `CONDITION_TYPE` | The Node Readiness condition written by the reporter, e.g. `projectcalico.org/CalicoReady`. | (required) |
```

Author: Updated

| `CHECK_INTERVAL` | How often the reporter polls `CHECK_ENDPOINT`. | `30s` |

> | `HEARTBEAT_PERIOD` | The maximum time the reporter can go without writing to the Node's condition, even when the observed health status hasn't changed. See [Reducing unnecessary node status writes](../user-guide/concepts.md#reducing-unnecessary-node-status-writes) for details. Accepts a Go duration string (`30s`, `2m`, `1h`). Invalid values are logged and fall back to the default. | `5m` |

Reviewer: ```suggestion
| `HEARTBEAT_PERIOD` | The maximum time the reporter can go without writing to the Node's condition if component health hasn't changed. See [Reducing unnecessary node status writes](../user-guide/concepts.md#reducing-unnecessary-node-status-writes). Accepts Go duration strings like `30s`, `2m`, `1h`. Invalid values are logged and fall back to the default. | `5m` |
```

Author: Updated

> | `IMPERSONATE_NODE` | When set to `"true"`, the reporter sends `Impersonate-User: system:node:<nodeName>` headers on every request, enabling the constrained impersonation authorization flow. See [Security](../operations/security.md#reporter-configuration) for details. | unset (uses the reporter's own ServiceAccount identity) |

Reviewer: Mention required k8s version / feature dependencies.

Author: Added Kubernetes version/feature gate for `IMPERSONATE_NODE` 
Added a bit of detail that it requires Kubernetes v1.35+ for KEP-5284 (Constrained Impersonation)

# Reporter Configuration

The `readiness-condition-reporter` is one of the ways Node Readiness Controller

> supports reporting node conditions from the node itself.

Reviewer: ```suggestion
supports reporting node conditions from the node.
```

Author: updated

This page lists configuration variables for deploying the
`readiness-condition-reporter`. These can be configured through environment
variables on its container. See the [CNI Readiness](../examples/cni-readiness.md)
and [Security Agent](../examples/security-agent-readiness.md) for examples.

## Environment Variables

| Variable | Description | Default |
| --- | --- | --- |

> | `NODE_NAME` | The name of the Node this reporter instance updates. Typically set via the Downward API (`fieldRef: spec.nodeName`) rather than a literal value. | (required) |

Reviewer: ```suggestion
| `NODE_NAME` | The host name of the underlying Node. The reporter uses it to identify the Node object in the API server for the readiness updates. Typically set via the Downward API (`fieldRef: spec.nodeName`). | (required) |
```

Author: updated

> | `CHECK_ENDPOINT` | The local HTTP endpoint the reporter polls to determine health, e.g. `http://localhost:9099/readiness`. | (required) |

Reviewer: ```suggestion
| `CHECK_ENDPOINT` | The HTTP endpoint the reporter polls to determine component health, e.g. `http://localhost:9099/healthz`. | (required) |
```

Author: updated

| `CONDITION_TYPE` | The Node Readiness condition written by the reporter, e.g. `projectcalico.org/CalicoReady`. | (required) |
| `CHECK_INTERVAL` | How often the reporter polls `CHECK_ENDPOINT`. | `30s` |
| `HEARTBEAT_PERIOD` | The maximum time the reporter can go without writing to the Node's condition if component health hasn't changed. See [Optimizing node status writes](../user-guide/concepts.md#optimizing-node-status-writes). Accepts Go duration strings like `30s`, `2m`, `1h`. Invalid values are logged and fall back to the default. | `5m` |

> | `IMPERSONATE_NODE` | When set to `"true"`, the reporter sends `Impersonate-User: system:node:<nodeName>` headers on every request, enabling the constrained impersonation authorization flow. Requires Kubernetes **v1.35+** for [KEP-5284](https://github.com/kubernetes/enhancements/issues/5284) (Constrained Impersonation) to actually be enforced by the API server. See [Security](../operations/security.md#reporter-configuration) for details. | unset (uses the reporter's own ServiceAccount identity) |

Reviewer: ```suggestion
| `IMPERSONATE_NODE` | When set to `"true"`, the reporter sends `Impersonate-User: system:node:<nodeName>` headers on every request, enabling the constrained impersonation authorization flow. Requires Kubernetes **v1.35+** for [Constrained Impersonation](https://kubernetes.io/docs/reference/access-authn-authz/user-impersonation/#constrained-impersonation) feature. See [Security](../operations/security.md#reporter-configuration) for details. | unset (uses the reporter's own ServiceAccount identity) |
```

Author: updated

> [!NOTE]
> `NODE_NAME`, `CHECK_ENDPOINT`, and `CONDITION_TYPE` together define *what*
> the reporter checks and *where* it writes the result. `CHECK_INTERVAL` and
> `HEARTBEAT_PERIOD` control *how often* it checks and *how often* it is
> guaranteed to write, respectively i.e. these are two different clocks and not

> > the same setting.

Reviewer: nit: This note feels redundant to me. Does this help better explain the concepts behind the check interval and heartbeat?

Author: I agree, removed the note, since it was restating things that are explained elsewhere
